### ISSUE TITLE
Reduce use of globals

### ISSUE THREAD

**Issue by AJenbo** [diasurgical/devilutionX#2435]
The following globals are being referenced a lot where it could instead have been done by reference, changing this will make functions more dependable/testable.

- [ ] `Items`
- [ ] `Lights`
- [ ] `Missiles`
- [ ] `Objects`
- [ ] `QuestData`
- [ ] `Quests`
- [ ] `VisionList`
- [x] `Players`
- [x] `Monsters`
- [x] `Towners`

**Comment by PracticalMetal** [comment]
Hi, I would like to work on this, can you suggest me ways to get started?

**Comment by AJenbo** [comment]
A good place to start is to go over the functions in objects.cpp and for the ones that have multiple direct usages of `Objects`, instead get it by reference at the start of the function, for example `AddL1Door` which has 6 references to `Objects`:

```cpp
void AddL1Door(int i, Point position, _object_id objectType)
{
	Objects[i]._oDoorFlag = true;
	if (objectType == _object_id::OBJ_L1LDOOR) {
		Objects[i]._oVar1 = dPiece[position.x][position.y];
		Objects[i]._oVar2 = dPiece[position.x][position.y - 1];
	} else { //_object_id::OBJ_L1RDOOR
		Objects[i]._oVar1 = dPiece[position.x][position.y];
		Objects[i]._oVar2 = dPiece[position.x - 1][position.y];
	}
	Objects[i]._oVar4 = 0;
}
```

Instead we can put it in a variable called door (if the function doesn't deal with destinct type just call it `object`) and use in place of the global:
```cpp
void AddL1Door(int i, Point position, _object_id objectType)
{
	auto &door = Objects[i];
	
	door._oDoorFlag = true;
	if (objectType == _object_id::OBJ_L1LDOOR) {
		door._oVar1 = dPiece[position.x][position.y];
		door._oVar2 = dPiece[position.x][position.y - 1];
	} else { //_object_id::OBJ_L1RDOOR
		door._oVar1 = dPiece[position.x][position.y];
		door._oVar2 = dPiece[position.x - 1][position.y];
	}
	door._oVar4 = 0;
}
```

At this point it clear to see that `i` is only used as an index in the global and we can instead take the object as a reference and push the look up to the caller:
```cpp
void AddL1Door(ObjectStruct &door, Point position, _object_id objectType)
{
	door._oDoorFlag = true;
	if (objectType == _object_id::OBJ_L1LDOOR) {
		door._oVar1 = dPiece[position.x][position.y];
		door._oVar2 = dPiece[position.x][position.y - 1];
	} else { //_object_id::OBJ_L1RDOOR
		door._oVar1 = dPiece[position.x][position.y];
		door._oVar2 = dPiece[position.x - 1][position.y];
	}
	door._oVar4 = 0;
}

void AddObject(_object_id objType, Point objPos)
{
	int oi = AvailableObjects[0];
	AddL1Door(Objects[oi], objPos, objType);
}
```

When doing this make sure to keep an eye out for functions that change the index value during the cause of the function, or track multiple objects.

And example of this is `Obj_Trap` which uses both `Objects[oti]` (`trigger`) and `Objects[i]` (`trap`), or `AddObjTraps` where the index (`oiTrap`) is assigned differently in two places.

**Comment by PracticalMetal** [comment]
Thanks for your reply I will be working on this.

**Comment by bcko** [comment]
I am interested in working on this as well. I will start working on `VisionList` first and open a pull request.

**Comment by AJenbo** [comment]
Nice 👍

**Comment by AJenbo** [comment]
I have unmarked `Players` as there are a lot of places where i could have used `&Players[MyPlayerId] == &player` instead of `MyPlayerId == pnum` which would help to allow a lot of function to take the player object rather then an index. The same might also be true in a few cases for `Monsters`. but it looks to be a lot harder to solve because of how they are tied to other objects and synced over the network.

Not that it might also be a good idea to crate a global for `PlayerStruct *MyPlayer` that we update at the same time as `MyPlayerId`, that way we can simply do `MyPlayer == &player`.

**Comment by Sid672** [comment]
@AJenbo Shall I start to update all the items now?

**Comment by AJenbo** [comment]
> @AJenbo Shall I start to update all the items now?

Yes if you could start working on `Objects[]` in objects.cpp that would be nice :)

**Comment by AJenbo** [comment]
@bcko I have added a global `Player *MyPlayer;` which can be used directly instead of `Players[MyPlayerId]`

**Comment by BCullen2201** [comment]
I'll get started on `Lights`

**Comment by BCullen2201** [comment]
#4478

**Comment by PlungedInCode** [comment]
Hi, Can i work on `Quests`?

**Comment by PlungedInCode** [comment]
Is there any unfinished modules where i can work?

**Comment by DakkJaniels** [comment]
Did you delete your message about working on Quests? I think any of those that are unchecked could still be worked on.

**Comment by PlungedInCode** [comment]
@DakkJaniels I checked `Quests` and find out that was refactored(I'm not sure though)
